crates/gitbutler-hunk-dependency/src/diff.rs

```diff
@@ -0,0 +1,76 @@
+use anyhow::{anyhow, Context};
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Diff {
+    pub old_start: i32,
+    pub old_lines: i32,
+    pub new_start: i32,
+    pub new_lines: i32,
+}
+
+impl Diff {
+    pub fn net_lines(&self) -> i32 {
+        self.new_lines - self.old_lines
+    }
+}
+
+fn count_context_lines<I, S>(iter: I) -> i32
+where
+    I: Iterator<Item = S>,
+    S: AsRef<str>,
+{
+    iter.take_while(|line| {
+        let line_ref = line.as_ref(); // Convert to &str
+        !line_ref.starts_with('-') && !line_ref.starts_with('+')
+    })
+    .fold(0i32, |acc, _| acc + 1)
+}
+
+impl TryFrom<String> for Diff {
```

Comment:
I believe you need to create these from BString - the reason being is that not all diffs are guaranteed to be utf8 serializable. For the UI we do create "diff for fronted" which is a utf8 (String) but those are lossy and are only used for display. Otherwise this may cause subtle / difficult to debug issues for some files and not others